pointcloud/merging.go

```diff
@@ -0,0 +1,144 @@
+package pointcloud
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"image/color"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"github.com/lucasb-eyer/go-colorful"
+	"go.opencensus.io/trace"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/spatialmath"
+)
+
+// CloudAndOffsetFunc is a function that returns a PointCloud with a pose that represents an offset to be applied to every point.
+type CloudAndOffsetFunc func(context context.Context) (PointCloud, spatialmath.Pose, error)
+
+// MergePointClouds takes a slice of points clouds with optional offsets and adds all their points to one point cloud.
+func MergePointClouds(ctx context.Context, cloudFuncs []CloudAndOffsetFunc, logger golog.Logger) (PointCloud, error) {
+	if len(cloudFuncs) == 0 {
+		return nil, errors.New("no point clouds to merge")
+	}
+	finalPoints := make(chan []PointAndData, 50)
+	activeReaders := int32(len(cloudFuncs))
+	for i, pcSrc := range cloudFuncs {
+		iCopy := i
+		pcSrcCopy := pcSrc
+		utils.PanicCapturingGo(func() {
+			_, span := trace.StartSpan(ctx, "pointcloud::MergePointClouds::Cloud"+fmt.Sprint(iCopy))
+			defer span.End()
+
+			defer func() {
+				atomic.AddInt32(&activeReaders, -1)
+			}()
+			pc, offset, err := pcSrcCopy(ctx)
+			if err != nil {
+				panic(err) // TODO(erh) is there something better to do?
+			}
+			var wg sync.WaitGroup
+			const numLoops = 8
+			for loop := 0; loop < numLoops; loop++ {
+				wg.Add(1)
+				f := func(loop int) {
+					defer wg.Done()
+					const batchSize = 500
+					batch := make([]PointAndData, 0, batchSize)
+					savedDualQuat := spatialmath.NewZeroPose()
+					pc.Iterate(numLoops, loop, func(p r3.Vector, d Data) bool {
+						if offset != nil {
+							spatialmath.ResetPoseDQTranslation(savedDualQuat, p)
+							newPose := spatialmath.Compose(offset, savedDualQuat)
+							p = newPose.Point()
+						}
+						batch = append(batch, PointAndData{P: p, D: d})
+						if len(batch) > batchSize {
+							finalPoints <- batch
+							batch = make([]PointAndData, 0, batchSize)
+						}
+						return true
+					})
+					finalPoints <- batch
+				}
+				loopCopy := loop
+				utils.PanicCapturingGo(func() { f(loopCopy) })
+			}
+			wg.Wait()
+		})
+	}
+	var pcTo PointCloud
+	var err error
+
+	dataLastTime := false // if there was data in the channel in the previous loop, continue reading.
```

Comment:
this may be a naive question and can be discussed later, but how fast is this code compared to a simpler version. This feels very complicated.

Reply:
I will make a ticket for simplifying this merging function.